generic `Client()` getters lazy loading (#2323)
zenml-io/zenml: docs/book/user-guide/advanced-guide/data-management/late-materialization.md

---
description: Always use up-to-date data in ZenML pipelines.
---

> # Late materialization in pipelines

Author: It is a kinda, of standard naming in the data world: https://docs.cloudera.com/cdw-runtime/cloud/impala-reference/topics/impala-lazy-materialization.html
People say "late" or "lazy" normally, but never hear "delayed" or "JIT". JIT might stand, but delayed sound negative, like it should happen but didn't happen on time.

> # Late materialization in pipelines

Author: https://aws.amazon.com/about-aws/whats-new/2017/12/amazon-redshift-introduces-late-materialization-for-faster-query-processing/

> # Late materialization in pipelines

Reviewer: Ok then I would go with late. Sorry I didn't realise it was a technical term. I'll update my review to make it easier to merge in.

> # Late materialization in pipelines

Reviewer: Updated now.

Often ZenML pipeline steps consume artifacts produced by one another directly in the pipeline code, but there are scenarios where you need to pull external data into your steps. Such external data could be artifacts produced by non-ZenML codes: for those cases it is advised to use [ExternalArtifact](../../../user-guide/starter-guide/manage-artifacts.md#consuming-external-artifacts-within-a-pipeline), but what if we plan to exchange data created within other ZenML pipelines?

ZenML pipelines are first compiled and executed at later point. During compilation phase all function calls are executed and this data is fixed as step input parameters. This is way late materialization of dynamic objects, like data artifacts, is crucial. Without late materialization it would not ber possible to pass as step input not yet existing artifact or its' metadata, which is often the case in the multi pipeline setting.

> We identify two major use-cases for exchange artifacts between pipelines:

Reviewer: ```suggestion
We identify two major use cases for exchanging artifacts between pipelines:
```

* You semantically group your data products using [ZenML Models](model-management.md#linking-artifacts-to-models)
* You prefer to use [ZenML Client](../../reference/python-client.md#client-methods) to bring all the pieces together

> In the sections below we will dive deeper into these use-cases from pipelines point of view.

Reviewer: ```suggestion
In the sections below we will dive deeper into these use cases from the pipeline perspective.
```

## Use ZenML Models to exchange artifacts

> The ZenML Model is an entity introduced by The Model Control Plane feature. The Model Control Plane is how you manage your models through this unified interface. It allows you to combine the logic of your pipelines, artifacts and crucial business data along with the actual 'technical model'.

Reviewer: ```suggestion
The ZenML Model is an entity introduced together with the Model Control Plane. The Model Control Plane is how you manage your models through a unified interface. It allows you to combine the logic of your pipelines, artifacts and crucial business data along with the actual 'technical model'.
```

Maybe add a link to the docs for MCP?

> [ZenML Models documentation](model-management.md#linking-artifacts-to-models) describes in great details how you can link various artifacts produced within pipelines to the model and here we would focus more on the consumption aspect of it.

Reviewer: ```suggestion
Documentation for [ZenML Models](model-management.md#linking-artifacts-to-models) describes in great detail how you can link various artifacts produced within pipelines to the model. Here we will focus more on the part that relates to consumption.
```

> First let's have a look at the two pipeline project, where the first pipeline is running training logic and the second runs batch inference leveraging trained model artifact:

Reviewer: ```suggestion
First, let's have a look at a two-pipeline project, where the first pipeline is running training logic and the second runs batch inference leveraging trained model artifact(s):
```

```python
from typing_extensions import Annotated
from zenml import get_pipeline_context, pipeline, Model
from zenml.enums import ModelStages
import pandas as pd
from sklearn.base import ClassifierMixin

@step
def predict(
    model: ClassifierMixin,
    data: pd.DataFrame,
) -> Annotated[pd.Series, "predictions"]:
    predictions = pd.Series(model.predict(data))
    return predictions

@pipeline(
    model=Model(
        name="iris_classifier",
        # Using the production stage
        version=ModelStages.PRODUCTION,
    ),
)
def do_predictions():
    # model name and version are derived from pipeline context
    model = get_pipeline_context().model
    inference_data = load_data()
    predict(
        # Here, we load in the `trained_model` from a trainer step
        model=model.get_model_artifact("trained_model"),  
        data=inference_data,
    )

if __name__ == "__main__":
    do_predictions()
```

> In the example above we used `get_pipeline_context().model` property to acquire the model context pipeline is running in. During pipeline compilation this context will be not evaluated, because `Production` model version is not stable version name and another model version can become `Production` before it comes to actual step execution. Same applies to calls like `model.get_model_artifact("trained_model")` - it will get stored in step configuration for late materialization during step run only.

Reviewer: ```suggestion
In the example above we used `get_pipeline_context().model` property to acquire the model context in which the pipeline is running. During pipeline compilation this context will not yet have been evaluated, because `Production` model version is not a stable version name and another model version can become `Production` before it comes to the actual step execution. The same applies to calls like `model.get_model_artifact("trained_model")`; it will get stored in the step configuration for delayed materialization which will only happen during the step run itself.
```

> It is also possible to achieve the same using bare `Client` methods reworking pipeline code as follows:

Reviewer: ```suggestion
It is also possible to achieve the same using bare `Client` methods reworking the pipeline code as follows:
```

```python
from zenml.client import Client

@pipeline
def do_predictions():
    # model name and version are directly passed into client method
    model = Client().get_model_version("iris_classifier", ModelStages.PRODUCTION)
    inference_data = load_data()
    predict(
        # Here, we load in the `trained_model` from a trainer step
        model=model.get_model_artifact("trained_model"),  
        data=inference_data,
    )
```

> In this case evaluation of actual artifact will happen in delayed fashion on step run only.

Reviewer: ```suggestion
In this case the evaluation of the actual artifact will happen only when the step is actually running.
```

## Use client methods to exchange artifacts

> If you don't yet use The Model Control Plane you can still exchange data between pipelines in late materialization fashion. Let's rework `do_predictions` pipeline code once again as follows:

Reviewer: ```suggestion
If you don't yet use the Model Control Plane you can still exchange data between pipelines with delayed materialization. Let's rework the `do_predictions` pipeline code once again as follows:
```

from zenml.client import Client
from zenml.enums import ModelStages
import pandas as pd
from sklearn.base import ClassifierMixin

@step
def predict(
    model1: ClassifierMixin,
    model2: ClassifierMixin,
    model1_metric: float,
    model2_metric: float,
    data: pd.DataFrame,
) -> Annotated[pd.Series, "predictions"]:
    # compare which models performs better on the fly
    if model1_metric < model2_metric:
        predictions = pd.Series(model1.predict(data))
    else:
        predictions = pd.Series(model2.predict(data))
    return predictions

@pipeline
def do_predictions():
    # get specific artifact version
    model_42 = Client().get_artifact_version("trained_model", version="42")
    metric_42 = model_42.run_metadata["MSE"].value

    # get latest artifact version
    model_latest = Client().get_artifact_version("trained_model")
    metric_latest = model_latest.run_metadata["MSE"].value

    inference_data = load_data()
    predict(
        model1=model_42,  
        model2=model_latest,
        model1_metric=metric_42,
        model2_metric=metric_latest,
        data=inference_data,
    )

if __name__ == "__main__":
    do_predictions()
```

> Here we also enriched `predict` step logic with metric comparison by MSE metric, so predictions are done on the best model possible. As before, calls like `Client().get_artifact_version("trained_model", version="42")` or `model_latest.run_metadata["MSE"].value` are not evaluating actual objects behind them during pipeline compilation, rather do so only in step run. By doing so we ensure that latest version is actually latest at the moment and not the latest on the pipeline compilation time.

Reviewer: ```suggestion
Here we also enriched the `predict` step logic with a metric comparison by MSE metric, so predictions are done on the best possible model. As before, calls like `Client().get_artifact_version("trained_model", version="42")` or `model_latest.run_metadata["MSE"].value` are not evaluating the actual objects behind them at pipeline compilation time. Rather, they do so only at the point of step execution. By doing so we ensure that latest version is actually the latest at the moment and not just the latest at the point of pipeline compilation.
```
